scheduler/src/main/java/gov/nasa/jpl/aerie/scheduler/StepSimulation.java

```diff
@@ -0,0 +1,255 @@
+import java.util.PriorityQueue;
+import java.util.Set;
+
+public class StepSimulation {
+
+  private Duration curTime;
+  private SimulationEngine engine;
+  private LiveCells cells;
+  private TemporalEventSource timeline;
+  private MissionModel<?> missionModel;
+
+  //used for looking up durations
+  private Map<SerializedActivity, TaskId> actToId;
+
+  //boolean stating whether the simulation results should be regenerated
+  private boolean areLastSimResultsDirty;
+
+  //simulation results so far
+  private SimulationResults lastSimResults;
+
+  //should the simulation be resetted everytime an activity is simulated
+  private boolean resetEveryTime;
+
+  //map from activity start time to serialized activity and its name
+  private LinkedHashMap<Duration, Pair<SerializedActivity, String>> activitiesInserted;
+
+  /*when we really want to reset the whole simulation*/
+  public void clearActInserted(){
+    activitiesInserted.clear();
+  }
+
+  public StepSimulation(MissionModel<?> missionModel){
+    this.missionModel = missionModel;
+    resetEveryTime = false;
+    activitiesInserted = new LinkedHashMap<>();
+  }
+
+  public void resetEveryTime(){
+    resetEveryTime= true;
+  }
+
+  public void initSimulation(){
+    actToId = new HashMap<>();
+    lastSimResults = null;
+    this.engine = new SimulationEngine();
+    activitiesInserted.clear();
+
+      /* The top-level simulation timeline. */
+      this.timeline = new TemporalEventSource();
+      areLastSimResultsDirty = false;
+      this.cells = new LiveCells(timeline, missionModel.getInitialCells());
+      /* The current real time. */
+      curTime = Duration.ZERO;
+
+      // Begin tracking all resources.
+      for (final var entry : missionModel.getResources().entrySet()) {
+        final var name = entry.getKey();
+        final var resource = entry.getValue();
+        engine.trackResource(name, resource, curTime);
+      }
+
+      // Start daemon task(s) immediately, before anything else happens.
+      {
+        final var daemon = engine.initiateTaskFromSource(missionModel::getDaemon);
+        final var commit = engine.performJobs(Set.of(SimulationEngine.JobId.forTask(daemon)),
+                                              cells, curTime, Duration.MAX_VALUE, missionModel);
+        timeline.add(commit);
+      }
+  }
+
+  public void simulateActivity(SerializedActivity activity, Duration startTime, String name){
+    if(startTime.shorterThan(curTime) || resetEveryTime){
+      var toBeReinserted = new LinkedHashMap<Duration, Pair<SerializedActivity, String>>();
+      toBeReinserted.putAll(activitiesInserted);
+      initSimulation();
+      insertAllActsBefore(toBeReinserted, startTime);
+      simulateAct(activity, startTime,name);
+      insertAllActsAfter(toBeReinserted, startTime);
+    }
+    simulateAct(activity, startTime,name);
+  }
+
+  private void insertAllActsBefore(LinkedHashMap<Duration, Pair<SerializedActivity, String>> acts, Duration startTime){
```

Comment:
It needs to be ordered by key (we want chronological insertion). So `OrderedMap` to be less specific ?

Reply:
I think you can use `SortedMap` here, since that expects the map to be sorted by key. (`LinkedHashMap` is ordered by insertion order into the map, which I don't think is what you want)

Reply:
I think Adrien has the caller inserting in sorted order, so it happens to work out. But you're right, @mattdailis, the method should document somehow (either by argument type or by comment) that it expects the provided map to be iterable in sorted order.

Honestly, the map provided to this method, `activitiesInserted`, is never really used as a map at all -- the only operations performed on it are essentially `put`, `entrySet`, and `clear`. This could be replaced with a sorted list (replace `put` with sorted insert, `entrySet` with regular `iterator()`, and `clear` stays the same). Then both `insertAllActsBefore` and `insertAllActsAfter` can be lifted into the lone caller, which itself can be significantly simplified.

Reply:
In the revision I am about to push, I let the ControlTask take care of the ordering now for scheduling multiple activities as it is done in the SimulationDriver. All of this has disappeared.